WebAssembly support

ijc8 opened the pull request:
Hi! I'm interested in getting RTcmix scores to play in the browser. This PR introduces a new WASM build type (joining the other embedded targets like PD, MAXMSP, and IOS) that compiles to [WebAssembly](https://webassembly.org/), enabling RTcmix to run in the browser and play audio via AudioWorklet.

You can try out a basic demo [here](https://ijc8.me/rtcmix-demo/).

There is an important caveat: some Minc functions don't work due to function pointer casting. Legacy functions are called as if they have the signature `double m_foo(float p[], int nargs, double pp[])`, but some of them don't.
For example, `rand()` is defined with just `double m_rand()`, and so the function pointer cast (to the type `double (*legacy_return) (float *, int, double *)`) is an incompatible cast with undefined behavior. (The undefined behavior turns out to be benign on e.g. x86 and ARM, but not in WebAssembly.)
Thus, `irand()` works fine while `rand()` fails with the error "indirect call signature mismatch". (See [this page](https://emscripten.org/docs/porting/guidelines/function_pointer_issues.html) from the Emscripten docs for more details.)

I believe this could be addressed by updating the Minc functions to use the full signature (with some functions ignoring some arguments), or by using the Emscripten workaround `EMULATE_FUNCTION_POINTER_CASTS` (which has performance overhead but may be less work).

Anyway, I would be happy to keep working on this and address this issue, along with any others that come up in merging. But I figured I should ask if there's interest in merging this upstream first! Let me know what you think (or if it'd be better to discuss this on the mailing list rather than on this PR).

ijc8 committed:
Header and namespace tweaks.

ijc8 committed:
Build system tweaks; add WASM build type.

ijc8 committed:
More build tweaks.

ijc8 committed:
Browser-based WebAssembly demo.

ijc8 committed:
RTcmix-WASM: Now in stereo!

ijc8 committed:
Check in README.

ijc8 committed:
Detect when score is finished; display status.

johgibso commented:
Hi Ian,

I was just thinking about how we should do this today! And here you are! :)

Doug Scott is the best person to pronounce on the function pointer casting issue. Obviously we need to clean up some ancient stuff.

The web demo is very inspiring!

Unfortunately, my time is extremely limited right now to help with any of this, but I just wanted to say I'm in favor! We just need to get it integrated without breaking anything.

Maybe since this is about a PR, the discussion should stay here. Doesn't really matter to me, though, whether it's here or on the dev mailing list. Should not be on the user mailing list (yet).

John



> On Jan 17, 2022, at 4:22 PM, Ian Clester ***@***.***> wrote:
> 
> Hi! I'm interested in getting RTcmix scores to play in the browser. This PR introduces a new WASM build type (joining the other embedded targets like PD, MAXMSP, and IOS) that compiles to WebAssembly <https://webassembly.org/>, enabling RTcmix to run in the browser and play audio via AudioWorklet.
> 
> You can try out a basic demo here <https://ijc8.me/rtcmix-demo/>.
> 
> There is an important caveat: some Minc functions don't work due to function pointer casting. Legacy functions are called as if they have the signature double m_foo(float p[], int nargs, double pp[]), but some of them don't.
> For example, rand() is defined with just double m_rand(), and so the function pointer cast (to the type double (*legacy_return) (float *, int, double *)) is an incompatible cast with undefined behavior. (The undefined behavior turns out to be benign on e.g. x86 and ARM, but not in WebAssembly.)
> Thus, irand() works fine while rand() fails with the error "indirect call signature mismatch". (See this page <https://emscripten.org/docs/porting/guidelines/function_pointer_issues.html> from the Emscripten docs for more details.)
> 
> I believe this could be addressed by updating the Minc functions to use the full signature (with some functions ignoring some arguments), or by using the Emscripten workaround EMULATE_FUNCTION_POINTER_CASTS (which has performance overhead but may be less work).
> 
> Anyway, I would be happy to keep working on this and address this issue, along with any others that come up in merging. But I figured I should ask if there's interest in merging this upstream first! Let me know what you think (or if it'd be better to discuss this on the mailing list rather than on this PR).
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/RTcmix/RTcmix/pull/28 <https://github.com/RTcmix/RTcmix/pull/28>
> Commit Summary
> 
> ff2fba8 <https://github.com/RTcmix/RTcmix/pull/28/commits/ff2fba8546ea7da788cc65799081f822ddd4566c> Header and namespace tweaks.
> 33083e3 <https://github.com/RTcmix/RTcmix/pull/28/commits/33083e31ecb6fccbfcc9ce5239eb5fa48a827b82> Build system tweaks; add WASM build type.
> 56f1b17 <https://github.com/RTcmix/RTcmix/pull/28/commits/56f1b175bc63227906e6ef383b36943f1d8079b6> More build tweaks.
> b2267f5 <https://github.com/RTcmix/RTcmix/pull/28/commits/b2267f53e28c1a35387da97076b02d777a8851c8> Browser-based WebAssembly demo.
> b3c8e51 <https://github.com/RTcmix/RTcmix/pull/28/commits/b3c8e5180823c5caf4fc71edf55efbdad70cda4b> RTcmix-WASM: Now in stereo!
> d32a4a1 <https://github.com/RTcmix/RTcmix/pull/28/commits/d32a4a1e66d13cb7912a310950bbdb0e201fd50b> Check in README.
> 107c82c <https://github.com/RTcmix/RTcmix/pull/28/commits/107c82cfc0a5aebf1e8c327bd56bc9aa83f07091> Detect when score is finished; display status.
> File Changes (18 files <https://github.com/RTcmix/RTcmix/pull/28/files>)
> A apps/wasmtest/Makefile <https://github.com/RTcmix/RTcmix/pull/28/files#diff-975d929c1078ad3395c67e59d1e48ec36e2b694683266359a103b28326578194> (11)
> A apps/wasmtest/README <https://github.com/RTcmix/RTcmix/pull/28/files#diff-733e54167c6083ae37c355f20ef7bcddd582be89f282b4d927c38442f10350a1> (16)
> A apps/wasmtest/glue.c <https://github.com/RTcmix/RTcmix/pull/28/files#diff-50780e28450799c6227bb9cc34e0cb5924b205a4f4527ea493f6cae3e7e29e53> (58)
> A apps/wasmtest/glue.js <https://github.com/RTcmix/RTcmix/pull/28/files#diff-e5ac60723d60c22c39cd61467db1c99361be5b02ee147a9d29156910a733fda4> (1)
> A apps/wasmtest/glue.wasm <https://github.com/RTcmix/RTcmix/pull/28/files#diff-146b3e1ce324fd212949231b72a2f9780e94b295c8afec0bd122355691ba5bfd> (0)
> A apps/wasmtest/index.html <https://github.com/RTcmix/RTcmix/pull/28/files#diff-c7fc258dc8a29010bf7137d2040b122d63a7a2e1b18297fbfc9503ed7ce30459> (73)
> A apps/wasmtest/worklet.js <https://github.com/RTcmix/RTcmix/pull/28/files#diff-c71718775299edfc16d1dae0aac4251d8bf7a3993649a7991580cc578e59d16f> (52)
> M insts/base/CHAIN/CHAIN.cpp <https://github.com/RTcmix/RTcmix/pull/28/files#diff-dd73dd99f0cb2ceeb1ceb14059c3102d184318507b262ebdfa66b4a0b10386b8> (1)
> M makefile.conf.in <https://github.com/RTcmix/RTcmix/pull/28/files#diff-c35ff18473b86283a6aa11d839801519da243073e651f0ea89ea0becfd7d06e1> (16)
> M site.conf <https://github.com/RTcmix/RTcmix/pull/28/files#diff-6e3d0202cc979eb8a63e67bf7d6cce9120b4f6357de8ca577f82cd08fca55464> (9)
> M src/audio/AudioFileDevice.cpp <https://github.com/RTcmix/RTcmix/pull/28/files#diff-d308e09b454dd1dae3d60b57996883bc737d721ccb94289d5ece2976d8b59051> (2)
> M src/audio/ThreadedAudioDevice.cpp <https://github.com/RTcmix/RTcmix/pull/28/files#diff-b2d9c04a8c1d9c269f4f31588764e2d3498164340f960534fbf2efb55c454e7c> (5)
> M src/rtcmix/Makefile <https://github.com/RTcmix/RTcmix/pull/28/files#diff-981f83485dac7e7927e587ce9ed1608c02eb8274c10d247346b60c9a41934363> (1)
> M src/rtcmix/RTcmixMain.cpp <https://github.com/RTcmix/RTcmix/pull/28/files#diff-d4677caa1b6463f1e5efc7cb756b008ede6988e63b026654d5e86c4f3e1d9be3> (2)
> M src/rtcmix/iosstuff/IOSLIBOBJS <https://github.com/RTcmix/RTcmix/pull/28/files#diff-2b1d6d917f62b8b2d74d826a35834e15f9f6468ad57810d8acd0d15f250d413a> (5)
> M src/rtcmix/sockdefs.h <https://github.com/RTcmix/RTcmix/pull/28/files#diff-ce5ddb4e14e07a9e89ebbd00cd2a2394f3d58f9339e2eff417d2f500de71400e> (3)
> M src/rtcmix/sound.c <https://github.com/RTcmix/RTcmix/pull/28/files#diff-f4cc27e6cd7d8cea9d87ab8223565080875892f996beb82b5e49dc8b306b2723> (1)
> M src/rtcmix/table.cpp <https://github.com/RTcmix/RTcmix/pull/28/files#diff-3ea9ae647249748399576ff2463e648a3dbd4316e977f3eebcea7e99fda394bf> (1)
> Patch Links:
> 
> https://github.com/RTcmix/RTcmix/pull/28.patch <https://github.com/RTcmix/RTcmix/pull/28.patch>
> https://github.com/RTcmix/RTcmix/pull/28.diff <https://github.com/RTcmix/RTcmix/pull/28.diff>
> —
> Reply to this email directly, view it on GitHub <https://github.com/RTcmix/RTcmix/pull/28>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB3TC3HXHHNRE7PQZCEXVSTUWSCBPANCNFSM5MFTN33A>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you are subscribed to this thread.
>

dougscott commented:
The changes to the existing RTcmix code all look fine.  I also agree that we should revisit all the old legacy functions and update their signatures.  However, before we do that I would like to consider simplifying all those old calls to have the updated signature (double *pp, int n_args) which will eliminate the old floating-point copy of the p-fields.  I need to do some code examination to determine if this would break anything in a serious way.
Do the changes in this PR represent everything else that is necessary to make this new configuration work?

ijc8 commented:
@johgibso

> I was just thinking about how we should do this today! And here you are! :)

Excellent! I originally worked on this a couple months ago, but it seems I was just in time with submitting the PR.

@dougscott 

> Do the changes in this PR represent everything else that is necessary to make this new configuration work?

Yes, everything should be there. The new build type is defined in `makefile.conf.in`, and the working example is in `apps/wasmtest` (named in analogy to the existing `apps/embedtest`).

`glue.js` and `glue.wasm` are build artifacts (included for convenience, but they could be omitted). `glue.c` is the actual source that gets linked against rtcmix_embedded, while `index.html` and `worklet.js` are helpers for the demo.
As the README mentions, this assumes you already have the [Emscripten SDK](https://emscripten.org/docs/getting_started/downloads.html) set up, and you have to do `emmake make` instead of plain `make`.
(Also, the PR currently switches the build type in `site.conf` from STANDALONE to WASM, which should be reverted before merging.)

Let me know if you run into any build issues, or if there's anything I can do to help with the function signature changes.

johgibso commented:
Thanks, Ian. I just looked you up and discovered you're working with Jason Freeman at GA Tech in my hometown. Thanks for appearing suddenly, like an angel, with this gift. :)

Looking forward to trying this out some more. I wonder what the prospects are for sending input to a running patch, via makeconnection or Brad's inlet method for rtcmix~.

John


> On Jan 18, 2022, at 1:59 PM, Ian Clester ***@***.***> wrote:
> 
> ﻿
> @johgibso
> 
> I was just thinking about how we should do this today! And here you are! :)
> 
> Excellent! I originally worked on this a couple months ago, but it seems I was just in time with submitting the PR.
> 
> @dougscott
> 
> Do the changes in this PR represent everything else that is necessary to make this new configuration work?
> 
> Yes, everything should be there. The new build type is defined in makefile.conf.in, and the working example is in apps/wasmtest (named in analogy to the existing apps/embedtest).
> 
> glue.js and glue.wasm are build artifacts (included for convenience, but they could be omitted). glue.c is the actual source that gets linked against rtcmix_embedded, while index.html and worklet.js are helpers for the demo.
> As the README mentions, this assumes you already have the Emscripten SDK set up, and you have to do emmake make instead of plain make.
> (Also, the PR currently switches the build type in site.conf from STANDALONE to WASM, which should be reverted before merging.)
> 
> Let me know if you run into any build issues, or if there's anything I can do to help with the function signature changes.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you were mentioned.

ijc8 committed:
Correct Makefile targets.

ijc8 committed:
Enable live input via "inlet" connections in demo.

ijc8 commented:
> I just looked you up and discovered you're working with Jason Freeman at GA Tech in my hometown. Thanks for appearing suddenly, like an angel, with this gift. :)

Ah, I've been found out! Glad to have the chance to contribute.

> I wonder what the prospects are for sending input to a running patch, via makeconnection or Brad's inlet method for rtcmix~.

The prospects are good. I've updated the [demo](https://ijc8.me/rtcmix-demo) to give an example of how this might be done. The HTML inputs serve as `hslider`s connected to two inlets. When you fiddle with them, the JS calls an exported C function (`send()`) which calls `RTcmix_setPField`.

Using "inlet" connections from JS is pretty straightforward. It may also be feasible to implement the "mouse" and "midi" (via [Web MIDI](https://developer.mozilla.org/en-US/docs/Web/API/Web_MIDI_API)) connection types; not sure if these are trickier to include in an embedded build.

johgibso commented:
This is great! Very promising.

My web coding is pretty basic. What would it take to get continuous output from one of these sliders into rtcmix? Right now, of course, it sends the current value when you release the mouse button. But ideally we would want continuous, smoothed output, while dragging the slider (or whatever the widget is). Would also be awesome to see all this available from p5.js.

(I have a secret agenda, which is eventually to get my instructional Max apps out of Max and into a more easily updateable web format. Here's one of the apps: https://cecm.indiana.edu/361/app-acoustics1.html <https://cecm.indiana.edu/361/app-acoustics1.html>.)

We do have smoothing available with makefilter("smooth", ...), but all the makeconnection variants, except for "inlet," implement the same smoothing algorithm via a "lag" parameter. See the docs and code for makeconnection "mouse" and "datafile" here: http://rtcmix.org/reference/scorefile/makeconnection.php <http://rtcmix.org/reference/scorefile/makeconnection.php>. I'm not sure why "inlet" excludes this, because it could be useful there as well. If it were implemented now, it would have to be via an optional parameter at the end of the argument list, with no smoothing as the default. That would work well for a user API, because for other variants, lag comes right after default, and it could still do that here.

pfield = makeconnection("mouse", "axis", min, max, default, lag[, "prefix"[, "units"], precision])

pfield = makeconnection("midi", min, max, default, lag, chan, "type"[, "subtype"])

pfield = makeconnection("datafile", "filename", lag, [skiptime[, timefactor[, filerate, format, swap]]])

So it could be...

pfield = makeconnection("inlet", inlet_number, default[, lag])


John



> On Jan 30, 2022, at 11:29 AM, Ian Clester ***@***.***> wrote:
> 
> 
> I just looked you up and discovered you're working with Jason Freeman at GA Tech in my hometown. Thanks for appearing suddenly, like an angel, with this gift. :)
> 
> Ah, I've been found out! Glad to have the chance to contribute.
> 
> I wonder what the prospects are for sending input to a running patch, via makeconnection or Brad's inlet method for rtcmix~.
> 
> The prospects are good. I've updated the demo <https://ijc8.me/rtcmix-demo> to give an example of how this might be done. The HTML inputs serve as hsliders connected to two inlets. When you fiddle with them, the JS calls an exported C function (send()) which calls RTcmix_setPField.
> 
> Using "inlet" connections from JS is pretty straightforward. It may also be feasible to implement the "mouse" and "midi" (via Web MIDI <https://developer.mozilla.org/en-US/docs/Web/API/Web_MIDI_API>) connection types; not sure if these are trickier to include in an embedded build.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/RTcmix/RTcmix/pull/28#issuecomment-1025179442>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB3TC3FKIVKQJ7JUFQ33YALUYVROXANCNFSM5MFTN33A>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you were mentioned.
>

ijc8 committed:
Send slider values continuously during mouse drag.

ijc8 commented:
> What would it take to get continuous output from one of these sliders into rtcmix?

Not much; just swapping out the `onchange` event listener for `oninput`. (See updated demo.)
Should also be easy to send in data from p5.js sliders (or other HTML inputs, `<canvas>`, etc.).

> I have a secret agenda, which is eventually to get my instructional Max apps out of Max and into a more easily updateable web format.

A worthy goal. In that vein, have you seen [WebPd](https://github.com/sebpiq/WebPd) or [Purr Data on the web](https://purrdata.glitch.me/)?

Re. smoothing, that sounds reasonable to me. (Perhaps `lag` was excluded on the assumption that smoothing would be done in the host patch. It could be done in JS, but maybe it's better for the score to have control over this.)

ijc8 committed:
Update .gitignore & Makefile.

- Rebuild wasmtest if RTcmix lib has been modified.
- Ignore Wasm build artifacts and more generated headers.
- Remove vestigial references to `notetags.h`.

dougscott commented:
I just checked in a large change which updates all the function prototypes to be both consistent and stripped of their old floating point array arguments.  Please see my comments from 1/17/22.  Question: are there other changes in this PR that need to go in to get this all to work?  If so, please rebase against the newest and resubmit.  -DS

ijc8 committed:
Add missing includes and compiler flag

ijc8 committed:
Fix remaining function signatures

ijc8 committed:
Further Makefile tweaks

ijc8 committed:
Update WebAssembly example

ijc8 committed:
Rebuild WebAssembly example

ijc8 commented:
@dougscott Great! I've rebased and made further tweaks as necessary. The original issue with this PR ("indirect call signature mismatch" when calling e.g. `rand()`) is resolved, and I've updated the demo [here](https://ijc8.me/rtcmix-demo).

dougscott commented on makefile.conf.in:
```diff
@@ -65,8 +65,6 @@ ifeq ($(ARCH),LINUX)
   ifndef AUDIODRIVER
 	AUDIODRIVER=OSS
   endif
-  CC = gcc
```
I need a bit more annotation about changes like this.  Why the change, and where do these get set now?

ijc8 replied:
The general answer (for this and the other changes) is that it's whatever I found necessary get this to build with [Emscripten](https://emscripten.org/).

The reason for this particular change is that Emscripten is based on LLVM, not gcc, and furthermore uses its own wrapper for clang (emcc). And in any case I didn't see a reason to insist on gcc on Linux; the user might prefer clang for standalone builds.

Unsetting `CC` allows an externally-provided `CC` to take precedence. Emscripten sets this to the path to `emcc` (via `emconfigure` and `emmake` as in apps/wasmtest/README; see [here](https://emscripten.org/docs/compiling/Building-Projects.html) and [here](https://github.com/emscripten-core/emscripten/blob/main/tools/building.py) for more information). If it's _not_ set externally, it defaults to `cc`, which seems like a reasonable default on Linux.

We could use `?=` if we want to specify `gcc` as the default while still allowing an external `CC` to override it. But I'm not sure why all of these variables need to be manually specified for each architecture in the first place; my impression is that `./configure` is supposed to determine sane values for `CC`, `LD`, etc. for "normal" builds.

dougscott commented on site.conf:
```diff
@@ -12,25 +12,26 @@
 #BUILDTYPE = MAXMSP
 #BUILDTYPE = PD
 #BUILDTYPE = IOS
+BUILDTYPE = WASM
```
Please dont check this in with WASM set as the default - confusing for others.

dougscott commented on src/audio/ThreadedAudioDevice.cpp:
```diff
@@ -61,10 +61,13 @@ int ThreadedAudioDevice::startThread()
 	pthread_attr_init(&attr);
-	int status = pthread_attr_setschedpolicy(&attr, SCHED_RR);
+	int status;
+#ifndef WASM
```
I dislike #ifdefs which are specific to a configuration like this -- can you explain why this has to be #ifdef'd here?

ijc8 replied:
I think I was having a linking issue with this call (`pthread_attr_setschedpolicy`). I can't reproduce it with the version of Emscripten I have now, so I'll remove the `#ifndef`.

dougscott commented on src/rtcmix/Makefile:
```diff
@@ -214,7 +214,6 @@ ifeq ($(RTLIBTYPE), STATIC)
 		$(LIBPFBUSOBJS) $(LIBMINCOBJS) $(LIBOBJOBJS) $(LIBRTHEAPOBJS) \
 		$(LIBSNDOBJS) $(LIBSTKOBJS)
 	LOAD = $(AR) rv
-	RANLIB = ranlib
```
Did you verify that all other configurations (in makefile.conf.in) provide a macro value for RANLIB?

ijc8 replied:
I didn't at the time, but it looks like all of the architectures (Linux, macOS, SGI :upside_down_face:, FreeBSD) do provide a value for `RANLIB`.

dougscott commented on src/rtcmix/iosstuff/IOSLIBOBJS:
```diff
@@ -35,8 +35,9 @@ LIBGENOBJS = ../../genlib/allpass.o ../../genlib/allpole.o \
 LIBINLETOBJS = ../control/maxmsp/RTInletPField.o ../control/maxmsp/inletglue.o
 LIBPFBUSOBJS = ../control/pfbus/PFBusPField.o ../control/pfbus/pfbusglue.o
 LIBMINCOBJS = ../parser/minc/Node.o ../parser/minc/callextfunc.o \
-../parser/minc/minc.o ../parser/minc/error.o ../parser/minc/sym.o \
```
Hah!  Good catch -- but are you actually building for iOS?  This stuff is so out of date!

ijc8 replied:
No, definitely not. However, I am doing static builds, and `IOSLIBOBJS` is included by `src/rtcmix/Makefile` for all static builds:
https://github.com/RTcmix/RTcmix/blob/6ad551a8930b1a0b3304b10dee9e6a99bab5a659/src/rtcmix/Makefile#L209-L212
Maybe this file should be moved/renamed?

dougscott commented on makefile.conf.in:
```diff
@@ -352,6 +358,10 @@ ifeq ($(ARCH),FREEBSD)
 endif
 
+ifeq ($(BUILDTYPE), WASM)
+    SHARED_LDFLAGS = -Wl
```
SHARED_LDFLAGS is set based upon ARCH for most other configurations - if this is truly a unique addition for WASM (and not due to the ARCH), please move this inside the new block at line 284.

ijc8 replied:
If I recall correctly (and I might not, since it was two years ago), this started out in the block at line 284, but I moved it further down because the blocks from 326-361 overwrite `SHARED_LDFLAGS`. The actual goal here is just to avoid `-shared`, preferably without repeating a `BUILDTYPE` or `RTLIBTYPE` check within various architecture blocks. Perhaps there's a cleaner way to do it.

ijc8 committed:
Restore default build type

ijc8 committed:
Remove unnecessary `#ifndef`
